README.md
# access-api-clients

Kount Access API Clients for various languages.  Choose the language you are interested in to see the README.md file related to that release.




CHANGELOG.md
# 03-08-2016

*  Fixed broken 2.1.0 unit test(403 return error)
*  Updated Unit tests to expect JSON object (we were not validating properly before)

# 02-15-2016

*  Updated default API version to 2.1.0

# 04-09-2015

*  Added nodejs 1.0.0 library

# 03-30-2015

*  Added php 1.0.0 library



nodejs/kount-access-api/test/kount-access-test.js
var kount = require('../lib/kount-access-api.js');
var expect = require("chai").expect;
var nock = require('nock');

//Default vaules for tests
var sessionId = "VALID_SESSION_ID";
var merchantId = 123456;
var apiKey = 'SOME_VALID_KEY';
var badApiKey = 'BAD_KEY';
var serverName = 'server.com';
var version = '0210';
var oldVersion = '0103';
var badVersion = '0100';
var user = 'admin';
var password = 'password';
var expectedGetOutput = { 'some':'data' };

// Test getDevice()
describe("kount-access", function() {
    describe(".getDevice", function() {
        it("should return a JSON array and no errors", function(done) {
            var mockApiServer = nock('https://server.com')
            .get('/api/device?s=' + sessionId + '&v=0210')
            .reply(200, expectedGetOutput);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDevice(sessionId, function(err,data){
              expect(err).equal(null);
              expect(data).eql(expectedGetOutput);
              done();
            });
        });
        it("should catch the missing sessionId and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDevice("", function(err,data){
              expect(err).is.equal("Invalid sessionId []");
              done();
            });
        });
        it("should catch the 403 and report up the error in the callback", function(done) {
            var mockApiServer = nock('https://server.com')
            .get('/api/device?s=' + sessionId + '&v=0210')
            .reply(403);
            var ka = new kount(merchantId, badApiKey, serverName);
            var results = ka.getDevice(sessionId, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:403");
              done();
            });
        });
        it("should return an HTML Help page", function(done) {
            var mockApiServer = nock('https://server.com')
            .get('/api/device?help=true&v=' + version)
            .reply(200, "<html><head></head><body>help page</body></html>");
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDeviceHelp(version, function(err,data){
              expect(err).is.equal(null);
              expect(data);
              done();
            });
        });
        it("should throw an error(400) when an invalid version is passed", function(done) {
            var mockApiServer = nock('https://server.com')
            .get('/api/device?help=true&v=' + badVersion)
            .reply(400);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDeviceHelp(badVersion, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:400");
              done();
            });
        });
    });
});

//Test getVelocity()
describe("kount-access", function() {
    describe(".getVelocity", function() {
        it("should return a JSON array and no errors", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/velocity')
            .reply(200, expectedGetOutput);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocity(sessionId, user,password, function(err,data){
              expect(err).is.equal(null);
              expect(data).eql(expectedGetOutput);
              done();
            });
        });
        it("should return a JSON array and no errors with an optional version", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/velocity')
            .reply(200, expectedGetOutput);
            var ka = new kount(merchantId, apiKey, serverName, oldVersion);
            var results = ka.getVelocity(sessionId, user,password, function(err,data){
              expect(err).is.equal(null);
              expect(data);
              done();
            });
        });
        it("should catch the missing sessionId and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocity(null, user, password, function(err,data){
              expect(err).is.equal("Invalid sessionId [null]");
              done();
            });
        });
        it("should catch the missing user name and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocity(sessionId, "", password, function(err,data){
              expect(err).is.equal("Missing user");
              done();
            });
        });
        it("should catch the missing password and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocity(sessionId, user, null, function(err,data){
              expect(err).is.equal("Missing password");
              done();
            });
        });
        it("should catch the 403 and report up the error in the callback", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/velocity')
            .reply(403);
            var ka = new kount(merchantId, badApiKey, serverName);
            var results = ka.getVelocity(sessionId, user, password, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:403");
              done();
            });
        });
        it("should return an HTML Help page", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/velocity', {help: "true", v: version})
            .reply(200, "<html><head></head><body>help page</body></html>");
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocityHelp(version, function(err,data){
              expect(err).is.equal(null);
              expect(data);
              done();
            });
        });
        it("should throw an error(400) when an invalid version is passed", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/velocity', {help: "true", v: badVersion})
            .reply(400);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getVelocityHelp(badVersion, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:400");
              done();
            });
        });
    });
});

//Test getDecision()
describe("kount-access", function() {
    describe(".getDecision", function() {
        it("should return a JSON array with decision and device and velcity information", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/decision')
            .reply(200, expectedGetOutput);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecision(sessionId, user, password, function(err,data){
              expect(err).is.equal(null);
              expect(data).eql(expectedGetOutput);
              done();
            });
        });
        it("should catch the missing sessionId and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecision(null, user, password, function(err,data){
              expect(err).is.equal("Invalid sessionId [null]");
              done();
            });
        });
        it("should catch the missing user name and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecision(sessionId, "", password, function(err,data){
              expect(err).is.equal("Missing user");
              done();
            });
        });
        it("should catch the missing password and report up the error in the callback", function(done) {
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecision(sessionId, user, null, function(err,data){
              expect(err).is.equal("Missing password");
              done();
            });
        });
        it("should catch the 403 and report up the error in the callback", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/decision')
            .reply(403);
            var ka = new kount(merchantId, "BAD_KEY", serverName);
            var results = ka.getDecision(sessionId, user, password, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:403");
              done();
            });
        });
        it("should return an HTML Help page", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/decision', {help: "true", v: version})
            .reply(200, "<html><head></head><body>help page</body></html>");
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecisionHelp(version, function(err,data){
              expect(err).is.equal(null);
              expect(data);
              done();
            });
        });
        it("should throw an error(400) when an invalid version is passed", function(done) {
            var mockApiServer = nock('https://server.com')
            .post('/api/decision', {help: "true", v: badVersion})
            .reply(400);
            var ka = new kount(merchantId, apiKey, serverName);
            var results = ka.getDecisionHelp(badVersion, function(err,data){
              expect(err).is.equal("HTTP ERROR STATUS:400");
              done();
            });
        });
    });
});




php/lib/kount_access_service.php
<?php
/**
 * Service API access class.
 * This class provides helper functions to utilize the Kount Access API
 * service.  In order to use this service you will be required to furnish:
 *   * Your Merchant ID
 *   * Your API Key
 *   * The API server you want to connect to
 *   * Information related to the queries you wish to make, including:
 *       * Session Id
 *       * username
 *       * password
 *
 * See the reference_implementation.php for sample usage.
 *
 * @version 2.0.0
 * @copyright 2015 Kount, Inc. All Rights Reserved.
 */
class Kount_Access_Service {

    /**
     * Version number
     */
    private $__version;

    /**
     * Private Kount Access API servername
     */
    private $__server_name;

    /**
     * Options for curl call
     */
    private $__encoded_credentials;

    /**
     * Constructor
     *
     * @param int merchant_id The Merchant's ID
     * @param string api_key API key assigned to merchant
     * @param string server_name The DNS name for the Kount Access API Server
     * @param string version The version of the API to access (0200 is the
     *                       default for this release of the SDK)
     * @return this
     */
    public function __construct ($merchant_id, $api_key, $server_name, $version = '0200') {
        $this->__server_name = $server_name;
        $this->__encoded_credentials = base64_encode($merchant_id  . ":" . $api_key);
        $this->__version = $version;
    } //end __construct

    /**
     * Gets the information for the Device based on the Session ID.
     *
     * @param string $session_id The Session ID used by the Device
     * @return array from JSON object decoded with details about the Device.
     */
    public function get_device ($session_id) {
      $endpoint = "https://$this->__server_name/api/device?v=$this->__version&s=$session_id";
        return $this->__call_endpoint($endpoint, "GET", null);
    } //end get_device


    /**
     * Gets the Velocity information for the Device based on the Session ID,
     * User ID and Password. These values are re-hashed in case the client
     * passed them in the clear prior to sending them off to the api server.
     *
     * @param string $session_id The Session ID used by the Device
     * @param string $user_id The user's User ID
     * @param string $password The user's Password
     * @return array from JSON object decoded with details about the Device.
     */
    public function get_velocity ($session_id, $user_id, $password) {
        $endpoint = "https://$this->__server_name/api/velocity";
        $u = hash('sha256', $user_id);
        $p = hash('sha256', $password);
        $a = hash('sha256', $user_id . ":" . $password);
        $data = array (
            s  => $session_id,
            v  => $this->__version,
            uh => $u,
            ph => $p,
            ah => $a
          );

        return $this->__call_endpoint($endpoint, "POST", $data);
    } //end get_velocity

    /**
     * Gets the Decision, Device information, and Velocity information for
     * the Device based on the Session ID, User ID and passwords. These values
     * are re-hashed in case the client passed them in the clear prior to
     * sending them off to the api server.
     *
     * @param string $session_id The Session ID used by the Device
     * @param string $user_id The user's User ID
     * @param string $password The user's Password
     * @return array from JSON object decoded with details about the Device.
     */
    public function get_decision ($session_id, $user_id, $password) {
        $endpoint = "https://$this->__server_name/api/decision";
        $u = hash('sha256', $user_id);
        $p = hash('sha256', $password);
        $a = hash('sha256', $user_id . ":" . $password);
        $data = array (
            s => $session_id,
            v  => $this->__version,
            uh => $u,
            ph => $p,
            ah => $a
          );
        return $this->__call_endpoint($endpoint, "POST", $data);
    } //end get_decision

    /**
     * Gets Help information for the get_decision endpoint by version.
     *
     * @param string version Version to get help for (defaults to 0200)
     * @return string HTML Help information, or a JSON array with error information
     */
    public function get_decision_help ($version = '0200') {
        $version = urlencode($version);
        $endpoint = "https://$this->__server_name/api/decision";
        $data = array (
            v => $version,
            help => "true",
          );
        return $this->__call_endpoint($endpoint, 'POST', $data, true);
    }

    /**
     * Gets Help information for the get_device endpoint by version.
     *
     * @param string version Version to get help for (defaults to 0200)
     * @return string HTML Help information, or a JSON array with error information
     */
    public function get_device_help ($version = '0200') {
        $version = urlencode($version);
        $endpoint = "https://$this->__server_name/api/device?v=$version&help=true";
        return $this->__call_endpoint($endpoint, 'GET', null, true);
    }

    /**
     * Gets Help information for the get_device endpoint by version.
     *
     * @param string version Version to get help for (defaults to 0200)
     * @return string HTML help information, or a JSON array with error information
     */
    public function get_velocity_help ($version = '0200') {
        $version = urlencode($version);
        $endpoint = "https://$this->__server_name/api/velocity";
        $data = array (
            v => $version,
            help => "true",
          );
        return $this->__call_endpoint($endpoint, 'POST', $data, true);
    }

    /**
     * Call a service endpoint.
     *
     * @param string $endpoint URL to endpoint
     * @param string $method Either POST or GET
     * @param array $params POST parameters
     * @return array JSON Response decoded or error array with cURL's
     *               ERROR_CODE and ERROR_MESSAGE values
     */
    private function __call_endpoint ($endpoint, $method=null, $params=null, $help=null) {
        $options = array(
            CURLOPT_FAILONERROR       => false,
            CURLOPT_USERAGENT         =>
                'Mozilla/5.0 (compatible; Service_Client/$Revision: 22622 $;)',
            CURLOPT_COOKIESESSION     => true,
            CURLOPT_HTTPHEADER        => array(
                "Accept: text/json",
                "Authorization: Basic $this->__encoded_credentials",
              ),
            CURLOPT_CONNECTTIMEOUT    => 3, // 3 seconds
            CURLOPT_TIMEOUT           => 5, // 5 seconds
            CURLOPT_HEADER            => true,
            CURLOPT_RETURNTRANSFER    => true,
            CURLOPT_URL               => $endpoint,
          );
        if ("POST" == $method) {
            $options[CURLOPT_POST] = true;
            $options[CURLOPT_POSTFIELDS] = $params;
        } else {
            $options[CURLOPT_CUSTOMREQUEST] = $method;
        }

        // Create curl resource
        $ch = curl_init();
        //Set Curl options
        curl_setopt_array($ch, $options);
        // Execute the request
        $raw_resp = curl_exec($ch);

        //Parse the response
        $err_code = curl_errno($ch);
        $resp_body = "";

        if (CURLE_OK == $err_code) {
            // parse the raw response
            $info = curl_getinfo($ch);
            $resp_code = (int) $info['http_code'];
            $hdr_size = $info['header_size'];
            $msg = mb_substr($raw_resp, $hdr_size,
                mb_strlen($raw_resp, 'latin1'), 'latin1');
            if (200 != $resp_code) {
                $resp_body = array(
                    ERROR_CODE    => $resp_code,
                    ERROR_MESSAGE => $msg,
                  );
            } else {
                if ($help) {
                    $resp_body = $msg;
                } else {
                    $resp_body =  json_decode($msg, true);
                }
            }
        } else {
            $resp_body = array (
                ERROR_CODE    => $err_code,
                ERROR_MESSAGE => curl_error($ch),
              );
        }

        curl_close($ch);
        return $resp_body;
    } //end call_endpoint
} //end kount_access_api



php/examples/reference_implementation.php
<?php
/**
 * This is a SIMPLE reference implementation of how to use the three API calls
 * from the kount_access_service.php class.
 * @copyright 2015 Kount, Inc. All Rights Reserved.
 */
require_once('./kount_access_service.php');

/**
 * The Kount_Access_Service can be set up once and then used multiple times
 * concurrently as needed. In this example we are creating it once and
 * passing it into all the example functions.  If you are just implementing 1
 * function, you can use this part and the function together. This just
 * makes the sample easier to manage as your credentials and servername are
 * in a single location.
 */
$main = function ($argc, $argv) {
    ///////////////////////////////////////////////////////////////////////////
    // Merchant information - replace with your own
    ///////////////////////////////////////////////////////////////////////////
    $merchant_id = 100100;
    $api_key = 'PUT_YOUR_API_KEY_HERE';

    ///////////////////////////////////////////////////////////////////////////
    // Kount Access API server to use
    ///////////////////////////////////////////////////////////////////////////
    $server = 'api-sandbox02.kountaccess.com';

    ///////////////////////////////////////////////////////////////////////////
    // Sample Data Section (update with data used in your testing)
    ///////////////////////////////////////////////////////////////////////////
    // Sample session ID (previously created by the server and passed to the
    // data collector when it ran on the login page)
    $session_id = '8f18a81cfb6e3179ece7138ac81019aa';

    // Users credentials used to login for the test
    $user = 'admin';
    $password ='password';

    ///////////////////////////////////////////////////////////////////////////
    // Now let's call each example and evaluate
    ///////////////////////////////////////////////////////////////////////////

    // Create an instance of the service
    $kount_access = new Kount_Access_Service($merchant_id, $api_key, $server);

    ///////////////////////////////////////////////////////////////////////////
    // If you are just looking for information about the device (like the
    // device id, or pierced IP Address) then use the get_device function.
    // This example shows how to get Device info and what it would look like as
    // an associative array
    ///////////////////////////////////////////////////////////////////////////
    echo "Example for calling kount_access->get_device('$session_id')\n";
    $response = $kount_access->get_device($session_id);
    evaluate_response($response);
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // If you make a bad request you will get an array abck with an ERROR_CODE
    // and ERROR_MESSAGE in it. This could be cURL related (Networking
    // issues?) or data releated (bad api key, invalid session_id, etc).
    // This example shows what a bad request's response would look like as
    // an associative array
    ///////////////////////////////////////////////////////////////////////////
    echo "Example for calling kount_access->get_device('BAD_SESSION_ID') with a bad session_id \n";
    $response = $kount_access->get_device('BAD_SESSION_ID');
    evaluate_response($response);
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // This is an example of the type of response to expect when requesting
    // Velocity information. The Device information is also included in this
    // response. You can use this Velocity information in your own decisioning
    // engine.
    ///////////////////////////////////////////////////////////////////////////
    echo "Example calling kount_access->get_velocity('$session_id', '$user', '$password')\n";
    $response = $kount_access->get_velocity($session_id, $user, $password);
    evaluate_response($response);
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // If you want Kount Access to evaluate possible threats using our
    // Thresholds Engine, you will want to call the get_decision endpoint.
    // This example shows how to make the decision call and what it would look
    // like as an associative array. This response includes Device information
    // and Velocity data in addition to the Decision information. The decision
    // value can be either "A" - Approve, or "D" - Decline.  In addition it will
    // show the ruleEvents evaluated that forced a "D" (Decline) result. If you
    // do not have any thresholds established it will always default to
    // "A" (Approve). For more information on setting up thresholds, consult the
    // User Guide.
    ///////////////////////////////////////////////////////////////////////////
    echo "Example calling kount_access->get_decision('$session_id', '$user', '$password')\n";
    $response = $kount_access->get_decision($session_id, $user, $password);
    evaluate_response($response);
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // If you want to know about a different version of the endpoints or what
    // the data within the json responses mean, you can call the help
    // functions in the service. The following three examples print out the
    // HTML help content from the server based on the version passed in.
    ///////////////////////////////////////////////////////////////////////////
    echo "Example of decision help: kount_access->get_decision_help()\n";
    print_help($kount_access->get_decision_help());
    wait_on_user();

    echo "Example of decision help: kount_access->get_device_help()\n";
    print_help($kount_access->get_device_help());
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // If you want to see a different versioni's help, then specify the version
    // in the input of the function
    ///////////////////////////////////////////////////////////////////////////
    echo "Example of passing a different version into help: kount_access->get_device_help('0103')\n";
    print_help($kount_access->get_velocity_help("0103"));
    wait_on_user();

    ///////////////////////////////////////////////////////////////////////////
    // If you specify an invalid version you will get a 400 server error
    ///////////////////////////////////////////////////////////////////////////
    echo "Example of passing a bad version into help: kount_access->get_device_help('ver 1.0')\n";
    evaluate_response($kount_access->get_device_help("ver 1.0"));
};

function print_help ($response) {
    if (is_array($response)) {
        evaluate_response($response);
    } else {
        echo "//////////////START RESPONSE//////////////////////\n";
        echo $response;
        echo "////////////////END RESPONSE//////////////////////\n";
    }
}

/**
 * Simple evaluator that either prints the errors, or the assoiative array
 * result
 */
function evaluate_response ($response) {
    echo "//////////////START RESPONSE//////////////////////\n";
    // Check for an error
    if ($response['ERROR_CODE']) {
        // Handle the Error. The two keys in the error array are ERROR_CODE and
        // ERROR_MESSAGE
        echo "Error code [" . $response['ERROR_CODE'] . "] returned with message [" .
          $response['ERROR_MESSAGE'] . "]\n";
    } else {
        // do something with the response
        echo "Got a response:";
        var_dump($response);
    }
    echo "///////////////END RESPONSE///////////////////////\n";
}

function wait_on_user () {
    echo "Press enter to continue\n";
    $handle = fopen ("php://stdin","r");
    $line = fgets($handle);
}

// kickoff main
$main($argc, $argv);


nodejs/kount-access-api/lib/kount-access-api.js
/**
 * Service API access class.
 * This class provides helper functions to utilize the Kount Access API
 * service.  In order to use this service you will be required to furnish:
 *   * Your Merchant ID
 *   * Your API Key
 *   * The API server you want to connect to
 *   * Information related to the queries you wish to make, including:
 *       * Session Id
 *       * username (hashed)
 *       * password (hashed)
 *
 * Online Help docucmentation for the endpoints can be retrieved from the
 * get*Help functions by passing in a version number.
 * See the reference_implementation.php for sample usage.
 *
 * @version 1.0.0
 * @copyright 2015 Kount, Inc. All Rights Reserved.
 */

/**
 * Required Libraries
 */
// needle is needed for http calls
var needle = require("needle");
// js-sha512 is used to hash account data prior to sending
var sha = require('js-sha512').sha512_256;


/**
 * Constructor.
 * MerchantId, apiKey, and serverName must be passed in. Version is optional.
 * @param int merchantId Your Merchant Id assigned by Kount
 * @param string apiKey API key for Kount Access
 * @param string serverName Server name (i.e. 'someserver.kountaccess.com')
 * provided by Kount for your environments (testing and production)
 * @param string version The Version of the API to use (defaults to "0210")
 */
module.exports = function (merchantId, apiKey, serverName, version) {
  this.merchantId = merchantId;
  this.apiKey = apiKey;
  this.serverName = serverName;
  this.defaultVersion = typeof version !== 'undefined' ? version : "0210";

  /**
   * Gets Device information based on the sessionId
   * @param string sessionId Session ID
   * @param function callback Expects a signature with 2 parameters (err, data)
   * The data object will contain a JSON Array object.
   */
  this.getDevice = function (sessionId, callback) {
    // validate required fields
    if (typeof sessionId !== 'string' || sessionId.length < 1) {
      return callback("Invalid sessionId [" + sessionId + "]",null);
    }
    var endpoint = "/api/device?s=" + sessionId + "&v=" + this.defaultVersion;
    this.makeRequest(endpoint, "GET", null, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    });
  } // end this.getDevice

  /**
   * Gets the velocity information for the device based on the sessionId,
   * user and passwords.  These values are re-hashed in case the client
   * passed them in the clear prior to sending them off to the api server.
   * @param string sessionId Session ID
   * @param string user The username pre-hashed. (We will re-hash it just
   * in case)
   * @param string password The password pre-hashed (we will re-hash it
   * just in case)
   * @param function callback Expects a signature with 2 parameters (err, data)
   * The data object will contain a JSON Array object.
   */
  this.getVelocity = function (sessionId, user, password, callback) {
    // validate required fields
    if (typeof sessionId !== 'string' || sessionId.length < 1) {
      return callback("Invalid sessionId [" + sessionId + "]",null);
    } else if (typeof user !== 'string' || user.length < 1) {
      return callback("Missing user",null);
    } else if (typeof password !== 'string' || password.length < 1) {
      return callback("Missing password",null);
    }

    // Setup request
    var endpoint = "/api/velocity";
    var parms = {
      s:sessionId,
      v: this.defaultVersion,
      uh: this.hash(user),
      ph: this.hash(password),
      ah: this.hash(user + ":" + password)
    };
    this.makeRequest(endpoint, "POST", parms, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    })
  } // end this.getVelocity

  /**
   * Gets the decision, device information, and velocity information for
   * the device based on the sessionId, user and passwords.  These values
   * are re-hashed in case the client passed them in the clear prior to
   * sending them off to the api server.
   * @param string sessionId Session ID
   * @param string user The username pre-hashed. (We will re-hash it just
   * in case)
   * @param string password The password pre-hashed (we will re-hash it
   * just in case)
   * @param function callback Expects a signature with 2 parameters (err, data)
   * The data object will contain a JSON Array object.
   */
  this.getDecision = function (sessionId, user, password, callback) {
    // validate required fields
    if (typeof sessionId !== 'string' || sessionId.length < 1) {
      return callback("Invalid sessionId [" + sessionId + "]",null);
    } else if (typeof user !== 'string' || user.length < 1) {
      return callback("Missing user",null);
    } else if (typeof password !== 'string' || password.length < 1) {
      return callback("Missing password",null);
    }

    // setup request
    var endpoint = "/api/decision";
    var parms = {
      s :sessionId,
      v : this.defaultVersion,
      ah: this.hash(user + ":" + password),
      uh: this.hash(user),
      ph: this.hash(password),
    };

    this.makeRequest(endpoint, "POST", parms, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    });
  } // end this.getDecision

  /**
   * Gets the help documentation as an HTML page for the version
   * specified.  If none are specified the default version is used.
   * @param string version The version to get docs for
   * @param function callback The Callback function to call (optionally)
   */
  this.getDeviceHelp = function (version, callback) {
    var endpoint = "/api/device?help=true&v=" + version;
    this.makeRequest(endpoint, "GET", null, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    });
  } //end getDecisionHelp

  /**
   * Gets the help documentation as an HTML page for the version
   * specified.  If none are specified the default version is used.
   * @param string version The version to get docs for
   * @param function callback The Callback function to call (optionally)
   */
  this.getVelocityHelp = function (version, callback) {
    var endpoint = "/api/velocity";
    var parms = {
      help: "true",
      v   : version,
    };
    this.makeRequest(endpoint, "POST", parms, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    });
  } //end getDecisionHelp



  /**
   * Gets the help documentation as an HTML page for the version
   * specified.  If none are specified the default version is used.
   * @param string version The version to get docs for
   * @param function callback The Callback function to call (optionally)
   */
  this.getDecisionHelp = function (version, callback) {
    var endpoint = "/api/decision";
    var parms = {
      help: "true",
      v   : version,
    };
    this.makeRequest(endpoint, "POST", parms, function(err, results) {
        if(err) {
          return callback(err, null);
        }
        callback(null, results);
    });
  } //end getDecisionHelp

  /**
   * hashing function that SHA-256 hashes the input and returns in the
   * callback
   * @param string text the string to hash
   * @return string Hashed value
   */
  this.hash = function (text) {
    return sha(text);
  } // end hash

  /**
   * Makes the https request
   *
   * @param string endpoint The endpoint on the server
   * @param string method Either GET or POST
   * @param string parms Data values for POST
   * @param function callback Function to call with a singature of (err,data)
   * in which data is the response.
   */
  this.makeRequest = function(endpoint, method, parms, callback) {

    var url = "https://" + this.serverName + endpoint;
    var options = {
       "username": this.merchantId,
       "password": this.apiKey,
       "accept": "text/json",
    };

    if (method == "GET") {
      var request = needle.get(url, options, function(err,response){
        if (!err && response.statusCode == 200) {
          callback(null,response.body)
        } else {
          if (err) {
            return callback(err,null)
          } else {
            return callback("HTTP ERROR STATUS:" + response.statusCode, null)
          }
        }
      });
    } else if (method == "POST") {
      var request = needle.post(url, parms, options, function(err,response){
        if (!err && response.statusCode == 200) {
          callback(null,response.body)
        } else {
          if (err) {
            return callback(err,null)
          } else {
            return callback("HTTP ERROR STATUS:" + response.statusCode, null)
          }
        }
      });
    }
  } // end makeRequest

} // end module.exports KountAccessApiService
